checking addon params

laurafitzgerald opened the pull request:
Removing WIP status

Addressed in this pr.
- Use GetOptions in cases where params have options listed - matches UI, the user is forced to select - set the first value in the array as the default.
- Check that val selected, or provided is contained in the options when there are options
- Move logic for install/edit get parameters section to function

tbrisker commented on cmd/install/addon/cmd.go:
```diff
@@ -197,17 +197,25 @@ func run(cmd *cobra.Command, argv []string) {
-					Help:     fmt.Sprintf("%s: %s", param.ID(), param.Description()),
-					Required: param.Required(),
+				options, ok := param.GetOptions()
+				if ok {
```
Should we use a `interactive.GetOption` prompt when there are predefined options for the parameter?

oriAdler commented on cmd/install/addon/cmd.go:
```diff
@@ -261,6 +269,12 @@ func run(cmd *cobra.Command, argv []string) {
 					}
 				}
+
+				if len(input.Options) > 0 && !contains(input.Options, val) {
```
We can use the `Contains` function from `pkg/helper/helpers.go`
```suggestion
				if len(input.Options) > 0 && !helper.Contains(input.Options, val) {
```

austincunningham commented:
/lgtm
verified the functionality 
checked out this branch 
ran
```bash
make install
```
On a rosa cluster installed the addon
```bash
rosa install addon --cluster=<cluster-name> managed-api-service
```
Got access to a menu to select the quota

![quota](https://user-images.githubusercontent.com/16667688/188851250-76746b45-856f-4d0c-8419-95e9fc3ffad9.gif)

The values for the quota don't reflect what the quota actually is `1==100k requests per day`, but we may address this elsewhere

tbrisker commented on cmd/edit/addon/cmd.go:
```diff
@@ -201,9 +158,13 @@ func run(cmd *cobra.Command, argv []string) {
 				}
 			}
+
+			if len(input.Options) > 0 && helper.Contains(input.Options, val) {
+				r.Reporter.Errorf("Expected %v to match one of the options /%v/", val, param.Options())
+				os.Exit(1)
+			}
```
This should probably be done in an input validator?

laurafitzgerald replied:
We want to validate in case of interactive input an parameter passing. The user has a limited set of options that can be passed that the addon can handle.

laurafitzgerald replied:
I searched in ocm-sdk-go addon_parameter_type. I didn't see a ValidateOptions function there unfortunately.

laurafitzgerald replied:
I found AddOnParameterOptionList here -> https://github.com/openshift-online/ocm-sdk-go/blob/main/clustersmgmt/v1/add_on_parameter_option_type.go#L191 Potentially should we be using this instead of this -> https://github.com/openshift-online/ocm-sdk-go/blob/main/clustersmgmt/v1/add_on_parameter_type.go#L46 ?

tbrisker commented on cmd/edit/addon/cmd.go:
```diff
@@ -129,69 +126,29 @@ func run(cmd *cobra.Command, argv []string) {
 
 		var val string
 		var hasVal bool
+		input := &interactive.Input{
```
using an input outside of interactive mode only to get the options seems strange to me

laurafitzgerald replied:
I'm declaring the object here in order to use it in the validation later. I can make a change where it just an empty object and put in the values inside the interactive block.

tbrisker commented on pkg/interactive/interactive.go:
```diff
@@ -46,7 +72,7 @@ func GetString(input Input) (a string, err error) {
 	}
 	question := input.Question
-	if !input.Required && dflt == "" {
+	if !input.Required && dflt == "" && !strings.Contains(strings.ToLower(question), "optional") {
```
This looks like something that should be fixed in the parameter description, not here (and below). That's what the `required` attribute is for.

laurafitzgerald replied:
@tbrisker unfortunately we are caught between the OCM UI and the ROSA Cli. We have the optional and required in the string of the description. So in ROSA CLI it's included twice. If we remove it from the addon parameter configuration we lose it in the UI.

laurafitzgerald replied:
Actually I think this might be fixable from our side. I've opened an MR [1] so will remove the code here.
[1] https://gitlab.cee.redhat.com/service/managed-tenants/-/merge_requests/3036

laurafitzgerald replied:
After merge of the above this is now working as expected so removed the code.

tbrisker replied:
🥳 glad you figured out a better way :)

tbrisker commented on cmd/install/addon/cmd.go:
```diff
@@ -197,61 +195,36 @@ func run(cmd *cobra.Command, argv []string) {
 				hasVal = true
 			}
-
+			if !hasVal {
+				return false
+			}
 			if hasVal {
```
you can get rid of the second `if` if you return early

tbrisker commented on cmd/install/addon/cmd.go:
```diff
@@ -197,61 +195,36 @@ func run(cmd *cobra.Command, argv []string) {
+				Help:     fmt.Sprintf("%s: %s", param.ID(), param.Description()),
+				Required: param.Required(),
+			}
+			options, ok := param.GetOptions()
+			if ok {
+				input.Options = []string{}
+				for _, opt := range options {
+					input.Options = append(input.Options, opt.Value())
+				}
+			}
```
is this part missing from the edit command on purpose?

laurafitzgerald replied:
No, not on purpose. I think I lost it when I refactored the function. Adding it in. 👍

tbrisker commented on cmd/install/addon/cmd.go:
```diff
@@ -197,61 +195,36 @@ func run(cmd *cobra.Command, argv []string) {
+				Help:     fmt.Sprintf("%s: %s", param.ID(), param.Description()),
+				Required: param.Required(),
+			}
+			options, ok := param.GetOptions()
+			if ok {
+				input.Options = []string{}
+				for _, opt := range options {
```
NP: if there are no options, `param.Options` will return a nil array, so you could use 
```suggestion
				for _, opt := range param.Options {
```
as a nil array will be a nop for range, and there is no need to initialize `input.Options` as it already has the zero value when the input is created.

tbrisker commented on pkg/interactive/interactive.go:
```diff
@@ -38,6 +40,30 @@ type Input struct {
+	if i == nil {
+		return 0
+	}
+	if i.Options == nil {
+		return 0
+	}
```
No need for this check, the length of a nil array is 0

tbrisker commented on pkg/interactive/interactive.go:
```diff
@@ -38,6 +40,30 @@ type Input struct {
 }
 
+func (i *Input) LenOptions() int {
+	if i == nil {
```
is this possible?

laurafitzgerald replied:
😵‍💫  it's been a while since writing some code. lol. hope you forgive the silly mistake. 😄

tbrisker commented on pkg/interactive/interactive.go:
```diff
@@ -38,6 +40,30 @@ type Input struct {
+	if i == nil {
+		return false
+	}
+	if i.Options == nil {
```
again, no need for this check - range over nil array is no-op

tbrisker commented on pkg/interactive/interactive.go:
```diff
@@ -38,6 +40,30 @@ type Input struct {
 	Validators []Validator
 }
 
+func (i *Input) LenOptions() int {
```
Not sure i understand why this method is needed instead of using `len(input.Options)` directly?

tbrisker commented on pkg/interactive/interactive.go:
```diff
@@ -312,3 +338,64 @@ func PrintHelp(help Help) error {
+			value = ""
+		}
+	case "number", "resource":
+		if input.LenOptions() > 0 {
+			value, err = GetOption(*input)
```
if we provide options, shouldn't that take precedence over the value type check and make it redundant?

laurafitzgerald replied:
I think there can be options in the case of a number of the different types? Is the suggestion here to move this above the case statement and ignore the case if we have options?

tbrisker replied:
yes, that's what i'm thinking - in case we have options we don't need to go through the case statement

tbrisker reviewed (commented):
thanks @laurafitzgerald, added some comments/questions inline

laurafitzgerald commented on pkg/interactive/interactive.go:
```diff
@@ -312,3 +338,64 @@ func PrintHelp(help Help) error {
 	return nil
 }
+
+func ProcessParamInteractive(addOnInstallationParam *cmv1.AddOnInstallationParameter, param *cmv1.AddOnParameter, input *Input) (error, string) {
```
@tbrisker I can move it elsewhere. I don't see any existing addon code I can add it to. The Input object is used a lot and the other function inside this file which is why I chose this. It's a refactor of logic which exists in both `cmd/install/addon/cmd.go` and `cmd/edit/addon/cmd.go` Didn't seem right to put it in either one of those and there was no addon specific spot in `cmd` that I could see. If you have a preference elsewhere I'm happy to move it. First pr to this repo so I hope you can forgive the lack of knowledge.

laurafitzgerald commented:
@tbrisker thanks for the review. I've addressed some comments and answered others. If you have time to take another look that would be really appreciated. 👍

tbrisker commented on pkg/interactive/interactive.go:
```diff
@@ -18,13 +18,12 @@ package interactive
-	"os"
-	"strconv"
-
 	"github.com/AlecAivazis/survey/v2"
 	"github.com/AlecAivazis/survey/v2/core"
 	"github.com/AlecAivazis/survey/v2/terminal"
+	"net"
+	"os"
+	"strconv"
```
This reorder is incorrect, i'm guessing some messed up editor config?

tbrisker commented on pkg/interactive/interactive.go:
```diff
@@ -176,6 +175,9 @@ func GetOption(input Input) (a string, err error) {
 	if !input.Required && dflt == "" {
 		question = fmt.Sprintf("%s (optional)", question)
 	}
+	if len(input.Options) > 0 {
+		dflt = input.Options[0]
+	}
```
what if we have a default defined for the input that isn't the first option?

tbrisker commented on pkg/interactive/addon.go:
```diff
@@ -0,0 +1,67 @@
+
+import (
+	"fmt"
+	cmv1 "github.com/openshift-online/ocm-sdk-go/clustersmgmt/v1"
```
NP: import order (this should be in a separate import section)

tbrisker commented on pkg/interactive/addon.go:
```diff
@@ -0,0 +1,67 @@
+
+func GetAddonParameter(addOnInstallationParam *cmv1.AddOnInstallationParameter, param *cmv1.AddOnParameter, input Input) (string, error) {
+	// Set default value based on existing parameter, otherwise use parameter default
+	dflt := param.DefaultValue()
+	if addOnInstallationParam != nil {
+		dflt = addOnInstallationParam.Value()
+	}
```
since we're only using `addOnInstallationParam` to override the default value when editing an addon installation, perhaps we could instead pass the default from the caller?

tbrisker commented on cmd/install/addon/cmd.go:
```diff
@@ -197,72 +196,53 @@ func run(cmd *cobra.Command, argv []string) {
 			var hasVal bool
+			var optionStrings []string
+
+			options, ok := param.GetOptions()
+			if ok {
+				for _, opt := range options {
+					optionStrings = append(optionStrings, opt.Value())
+				}
+			}
```
NP: I think we can safely get rid of the ok check here
```suggestion
			for _, opt := range param.Options() {
				optionStrings = append(optionStrings, opt.Value())
			}
```

tbrisker commented on cmd/install/addon/cmd.go:
```diff
@@ -197,72 +196,53 @@ func run(cmd *cobra.Command, argv []string) {
 				hasVal = true
 			}
+			if !hasVal {
+				return false
```
This will cause the parameter loop to stop iterating, i.e. if one parameter is missing a value, the rest of the parameters will be skipped. With `.Each` loops, `return false` is equal to `break` and `return true` to `continue`.

tbh i'm not sure we really need this `hasVal` flag, but that's outside the scope of this PR.

laurafitzgerald replied:
If I remove this return now that the logic is refactored, there isn't any use of the `hasVal` flag. I think it could be removed and the value of `val set to `""` on instantiation. I'm happy to do this in this pr if you agree?

ciaranRoche replied:
yes, please remove `hasVal` as it is not required any more

tbrisker commented on cmd/install/addon/cmd.go:
```diff
@@ -18,10 +18,9 @@ package addon
 import (
 	"fmt"
-	"net"
+	"github.com/openshift/rosa/pkg/helper"
```
np: import order

tbrisker commented on cmd/edit/addon/cmd.go:
```diff
@@ -18,19 +18,16 @@ package addon
 	"github.com/openshift/rosa/pkg/ocm"
 	"github.com/openshift/rosa/pkg/rosa"
+	"github.com/spf13/cobra"
+	"os"
+	"regexp"
+	"strings"
```
np: import order

tbrisker reviewed (commented):
Great progress @laurafitzgerald ! please see a few small comments inline

ciaranRoche commented on cmd/edit/addon/cmd.go:
```diff
@@ -18,19 +18,17 @@ package addon
 	"github.com/openshift/rosa/pkg/interactive"
 	"github.com/openshift/rosa/pkg/ocm"
 	"github.com/openshift/rosa/pkg/rosa"
+	"github.com/spf13/cobra"
```
np -> ordering of imports

Should be in this order
```
native imports

external imports

Internal imports
```

ciaranRoche commented on cmd/install/addon/cmd.go:
```diff
@@ -197,72 +196,51 @@ func run(cmd *cobra.Command, argv []string) {
+
+			options, _ := param.GetOptions()
+			for _, opt := range options {
+				optionStrings = append(optionStrings, opt.Value())
```
I think it would be better to create a list of `option.Name()`. The value is arbitrary, where as the name has some context behind it

laurafitzgerald commented:
@ciaranRoche implemented your suggestions. thanks for the review.

ciaranRoche commented:
JIRA - SDA-6845

ciaranRoche commented:
/lgtm
